Add in persistence to maintenance windows

Ktmi opened the pull request:
Fix #49 
Fix #48 

This PR seeks to resolve #49, implementing persistence to maintenance windows using mongo db. Upon adding a maintenance window, it will automatically be added to the database and scheduler. If kytos shuts down, then restarts, the napp should automatically attempt to re-dispatch any running maintenances, and reschedule any pending maintenance.

# Notes

This PR is dependent on kytos-ng/kytos#308, for making the database accessible through the controller. Also the requirements files may need to be updated in order to use that PR.

viniarck reviewed (commented):
Great progress @Ktmi. It's nice to see how this NApp is evolving and having persistence. 

I appreciate your contribution. I've just done a first review.

Ktmi commented on main.py:
```diff
@@ -3,16 +3,55 @@
 
 import pytz
+from attrs import evolve
+from cattrs import BaseValidationError, Converter
```
So the reason I selected `attrs` over `pydantic` comes down to issues I was having early on with the IDs. Originally I wanted to use the ID types we have in core, rather than the string representations. With `pydantic`, I was running into issues with the id type not already being a pydantic type due to their dual string/object nature. With `cattrs`, it seemed simpler to handle this special case, even being able to have separate strategies between the front-end and back-end.

Ultimately, the whole idea fell through, as the apis in other napps already use use the string representations rather than object representations for the ids. So I could convert back to pydantic.

viniarck replied:
@Ktmi, I see. Thanks for the information. Since the idea fell through, and although the IDs ended up extremely nicely in the core when building the objects, and since we're using list of string here, which also simplified and ended-up very nice in the API, let's switch to `pydantic` then all things considered.  Since we don't have a blocker here to use `pydantic` let's stick with it, it'll likely also be less cognitive load not to  learn a new equivalent-ish library and having to keep up with upstream updates for existing core developers and new contributors who might join our team.

viniarck commented on main.py:
```diff
@@ -45,118 +84,117 @@ def shutdown(self):
-            raise BadRequest('Start in the past not allowed')
-        if maintenance.end <= maintenance.start:
-            raise BadRequest('End before start not allowed')
+            maintenance = converter.structure(data, MW)
```
@Ktmi what's the expected behavior if a non active or non existing switch (interface or link) id is present in any of the lists? that might be outside of the scope of this PR, but it's something that we need to align either in the blueprint or a new issue. This can also lead to invalid data being inserted into the collection, which although is not desirable, it might be acceptable if we make clear that's not being validated in this iteration that will be delivered.

Ktmi replied:
So, Ive been thinking about this one, and I think that we should allow ids for things we can't confirm to exist to be used. My main reasoning for this is that when we create a maintenance window, we won't be 100% certain of the state of the network's topology. It could be that switches where removed or added, making some ids that were previously invalid into valid, or valid into invalid.

viniarck replied:
Right @Ktmi, I believe this would make sense for this iteration. It's probably something to be refined in the blueprint, just so network operators who will use in production are also aware of this, and then in the future we can keep evolving and even try to integrate more with `topology` as new requirements are discussed and aligned. Thanks.

italovalcy replied:
@viniarck and @Ktmi, the way I see is: we could have a parameter so that the operator informs Kytos that he/she is requesting an inexistent item on purpose and then the validation should be ignored. Something like `force` or `skip_check_exists` , etc.

italovalcy replied:
I've mapped Issue #65 to continue the discussion about this. Let's simplify for now and then we refine the requirement for the next release.

viniarck reviewed (changes requested):
Requested changes in the previous comments. I'll update the review status accordingly. Thanks, David.

viniarck commented on controllers/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    def add_window(self, window: MaintenanceWindow):
+        self.windows.insert_one(window.dict())
+
+    def get_window(self, mw_id: MaintenanceID) -> MaintenanceWindow:
```
Should the return type be `Optional[MaintenanceWindow]`? It seems that it can return `None`

viniarck reviewed (commented):
@Ktmi thanks for the updates, great how it's shaping up. I've just done another partial review based on the updates. If you need another one in the future let me know.

Ktmi commented on models.py:
```diff
@@ -3,228 +3,250 @@
+        """Begins executing the maintenance window
+        """
+        # Get Maintenance from DB
+        window = self.db.get_window(mw_id)
```
I did what I could for this, at least for starting and ending maintenance windows. For generalized updates, I still need to validate the data, and the best way to do so is get the object from the db, construct it in python, then validate it.

viniarck replied:
Right. Let's keep this way in this iteration.

Ktmi commented on controllers/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    def bootstrap_indexes(self) -> None:
+        """Bootstrap all maintenance related indexes."""
+        index_tuples = [
+            ("maintenance.windows", [("id", pymongo.ASCENDING)]),
```
Made the id index unique. Inserting into the DB will also check if there is already a window with the same ID, and if so reject the insert operation.

Ktmi commented on main.py:
```diff
@@ -3,16 +3,36 @@
+TIME_FMT = "%Y-%m-%dT%H:%M:%S%z"
+
+
+def validate_mw(maintenance: MW):
```
Validation is now handled by the validators on the MaintenanceWIndow model.

viniarck commented on ui/k-toolbar/main.kytos:
```diff
@@ -1,26 +1,36 @@
+        <k-button tooltip="Maintenance Windows"
+          title="List Maintenance Windows"
+          icon="desktop"
+          :on_click="list_windows">
```
David, if you could review the JS changes, in the UI it's hanging when listing existing windows, if you could fix and double check:

![20221219_115714](https://user-images.githubusercontent.com/1010796/208456015-39c56382-9882-4af7-84a1-96de9a6f73ff.png)

viniarck commented on controllers/__init__.py:
```diff
@@ -0,0 +1,126 @@
+            {'id': window.id},
+            {
+                '$setOnInsert': {
+                    **window.dict(exclude=['inserted_at', 'updated_at']),
```
```suggestion
                    **window.dict(exclude={'inserted_at', 'updated_at'}),
```

`pydantic` [exclude arg expects a type  `AbstractSetIntStr | MappingIntStrAny | None`](https://github.com/pydantic/pydantic/blob/main/pydantic/main.py#L199), as it is, it'll crash.

viniarck commented on requirements/dev.txt:
```diff
@@ -1,6 +1,6 @@
 #
-# This file is autogenerated by pip-compile with python 3.9
-# To update, run:
+# This file is autogenerated by pip-compile with Python 3.10
```
Thanks for reverting the requirements changes. It's great to have them regenerated with the recent updates from `kytos` transitive dependencies, however, to avoid confusion, could you regenerated with py `3.9` that's the supported version? The result will probably be the same, but to avoid surprises. Also if you could keep `typing-extensions` pinned as `typing-extensions>=4.0.1`

> Starting with version 4.0.0, typing_extensions uses [Semantic Versioning](https://semver.org/). The major version is incremented for all backwards-incompatible changes. Therefore, it's safe to depend on typing_extensions like this: typing_extensions >=x.y, <(x+1), where x.y is the first version that includes all features you need.

viniarck commented on CHANGELOG.rst:
```diff
@@ -7,11 +7,15 @@ All notable changes to the Maintenance NApp will be documented in this file.
 =======
-
 - Updated REST API endpoints to include version numbers. Current version is `\v1`.
+- Changed API to use descriptive strings for `status` of maintenance windows instead of integers.
```
Great enhancements, @Ktmi. If you could also update the [openapi.yml spec](https://github.com/kytos-ng/maintenance/blob/master/openapi.yml) to reflect the changes accordingly or if you want to handle that on another PR and issue that's supposed to land on this `2022.3` version is also OK.

viniarck commented on requirements/dev.in:
```diff
@@ -3,3 +3,4 @@
 -e .[dev]
 pytest-asyncio==0.18.3
 black==22.3.0
+cattrs==22.2.0
```
Requirements discussion continued here https://github.com/kytos-ng/maintenance/pull/64#discussion_r1052336192

viniarck commented:
@Ktmi , I'm doing partial reviews to try to avoid the PR to stall in review as I get notifications from a few commits being pushed, but if you prefer only a next review when you're done pushing the commits that also works, let me know if you have a preference.

Ktmi commented:
I think I'm going to need a little help understanding the javascript for edit window. I can get it to properly parse the new window structure, however, the request that it produces for updating the window ends up malformed. 

Here's what I have so far:
https://gist.github.com/Ktmi/69599c5bbf736b6a6baa19a9a8a17df4

Here's an example of a malformed request:

```json
{
	"description": "",
	"end": "2022-12-30T13:00:00+0000",
	"id": "Test topo",
	"interfaces": [
		{
			"description": "00:00:00:00:00:00:00:01:1",
			"value": "00:00:00:00:00:00:00:01:1"
		},
		{
			"description": "00:00:00:00:00:00:00:01:2",
			"value": "00:00:00:00:00:00:00:01:2"
		},
		{
			"description": "00:00:00:00:00:00:00:01:4294967294",
			"value": "00:00:00:00:00:00:00:01:4294967294"
		},
		{
			"description": "00:00:00:00:00:00:00:02:1",
			"value": "00:00:00:00:00:00:00:02:1"
		},
		{
			"description": "00:00:00:00:00:00:00:02:2",
			"value": "00:00:00:00:00:00:00:02:2"
		},
		{
			"description": "00:00:00:00:00:00:00:02:4294967294",
			"value": "00:00:00:00:00:00:00:02:4294967294"
		},
		"00:00:00:00:00:00:00:01:1",
		"00:00:00:00:00:00:00:01:2",
		"00:00:00:00:00:00:00:01:4294967294",
		"00:00:00:00:00:00:00:02:1",
		"00:00:00:00:00:00:00:02:2",
		"00:00:00:00:00:00:00:02:4294967294"
	],
	"links": [],
	"start": "2022-12-30T12:00:00+0000",
	"switches": [
		{
			"description": "00:00:00:00:00:00:00:01",
			"value": "00:00:00:00:00:00:00:01"
		},
		{
			"description": "00:00:00:00:00:00:00:02",
			"value": "00:00:00:00:00:00:00:02"
		},
		"00:00:00:00:00:00:00:01",
		"00:00:00:00:00:00:00:02"
	]
}
```

viniarck commented:
> I think I'm going to need a little help understanding the javascript for edit window. I can get it to properly parse the new window structure, however, the request that it produces for updating the window ends up malformed.
> 
> Here's what I have so far: https://gist.github.com/Ktmi/69599c5bbf736b6a6baa19a9a8a17df4
> 
> Here's an example of a malformed request:
> 
> ```json
> {
> 	"description": "",
> 	"end": "2022-12-30T13:00:00+0000",
> 	"id": "Test topo",
> 	"interfaces": [
> 		{
> 			"description": "00:00:00:00:00:00:00:01:1",
> 			"value": "00:00:00:00:00:00:00:01:1"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:01:2",
> 			"value": "00:00:00:00:00:00:00:01:2"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:01:4294967294",
> 			"value": "00:00:00:00:00:00:00:01:4294967294"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:02:1",
> 			"value": "00:00:00:00:00:00:00:02:1"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:02:2",
> 			"value": "00:00:00:00:00:00:00:02:2"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:02:4294967294",
> 			"value": "00:00:00:00:00:00:00:02:4294967294"
> 		},
> 		"00:00:00:00:00:00:00:01:1",
> 		"00:00:00:00:00:00:00:01:2",
> 		"00:00:00:00:00:00:00:01:4294967294",
> 		"00:00:00:00:00:00:00:02:1",
> 		"00:00:00:00:00:00:00:02:2",
> 		"00:00:00:00:00:00:00:02:4294967294"
> 	],
> 	"links": [],
> 	"start": "2022-12-30T12:00:00+0000",
> 	"switches": [
> 		{
> 			"description": "00:00:00:00:00:00:00:01",
> 			"value": "00:00:00:00:00:00:00:01"
> 		},
> 		{
> 			"description": "00:00:00:00:00:00:00:02",
> 			"value": "00:00:00:00:00:00:00:02"
> 		},
> 		"00:00:00:00:00:00:00:01",
> 		"00:00:00:00:00:00:00:02"
> 	]
> }
> ```

@Ktmi this patch is in the right direction. I suspect the issue is related how the `this.chosen_*` synced vars are being re-used in the request payload. It used to build another list `jsonItems` based on the chosen items. 

Notice here for instance, that if you don't select anything it'll send the chosen contents which will result in an invalid payload:

![20221220_164122](https://user-images.githubusercontent.com/1010796/208754289-f8ababa8-2e16-4424-b1b6-9b35da4a9610.png)

If you were simulate JS click events to update the content of the chosen items it would succeed:

![20221220_164212](https://user-images.githubusercontent.com/1010796/208754586-f7c65089-70f9-4936-b894-77efce2c08ea.png)

Ktmi commented:
I think at this point most of the issues with this PR should be resolved. @viniarck could you please do a final review and make sure I properly addressed your concerns?

viniarck commented on ui/k-info-panel/edit_window.kytos:
```diff
@@ -127,57 +127,39 @@
+                
                 // For every chosen link
-                for(i = 0; i < this.chosen_links.length; i++) {
+                for(item of this.chosen_links) {
```
@Ktmi, if you were to try to edit and save the name of the MW, when the expected tz format is resulting an error, could you look into it?

![20221222_135703](https://user-images.githubusercontent.com/1010796/209187095-5bd7fef4-39d0-4108-9138-d65921659047.png)

italovalcy replied:
I'm not sure if the date format issue was fixed, but I'm facing another issue when editting the table:

![Screenshot 2023-01-13 at 10 29 26 AM](https://user-images.githubusercontent.com/168159/212331354-79fab045-ed64-4247-ae46-c62d8668645f.png)

I'm using the most updated version of the UI (including the changes from PR https://github.com/kytos-ng/ui/pull/33)

italovalcy replied:
This issue was fixed in the last commits. The issue I was having is related to https://github.com/kytos-ng/ui/issues/34

viniarck commented on controllers/__init__.py:
```diff
@@ -0,0 +1,126 @@
+    before_sleep=before_sleep,
+    retry=retry_if_exception_type((AutoReconnect,)),
+)
+class MaintenanceController:
```
@Ktmi this module and class hasn't been covered in unit tests, the total coverage fell to 62%, we're trying to maintain at least 85%. 

Currently, we've been unit testing DB controllers this way, examples below:

https://github.com/kytos-ng/flow_manager/blob/master/tests/unit/test_flow_controller.py
https://github.com/kytos-ng/topology/blob/master/tests/unit/test_topo_controller.py

italovalcy replied:
With the latest commits, this pendency was fixed. According to [Scrutinizer's last execution](https://scrutinizer-ci.com/g/kytos-ng/maintenance/inspections/6c472997-3965-4cf2-81e0-a1e65329d306), code coverage increased to 87%! Very nice job, @Ktmi!

viniarck reviewed (commented):
> I think at this point most of the issues with this PR should be resolved. @viniarck could you please do a final review and make sure I properly addressed your concerns?

Sure, @Ktmi. I've closed all of the prior threads (except one that is related to blueprint, feel free to close it in the future). 

This PR is almost there, thanks for pushing these commits. It's  great how this PR is turning out. I've just done another pass and I opened new code review threads. 

1 - Let us know if you had the opportunity to explore locally this change, especially also with the UI that we don't have test coverage for it at the moment. 
2 - Could you run e2e tests with this branch to confirm it's still passing?
3 - Could you also fix the linter errors?

Ktmi commented on main.py:
```diff
@@ -45,118 +47,118 @@ def shutdown(self):
+    def get_all_mw(self):
+        """Return all maintenance windows."""
+        maintenances = self.scheduler.list_maintenances()
+        return current_app.response_class(
```
I used pydantics jsonfiy specifically to match `TIME_FMT` we are using. If we remove the requirement to match `TIME_FMT`, then we can perform marshalling with flask. This should also fix the issue with submitting times that don't include tz info.

viniarck replied:
I see. So, this won't be a blocker for this PR, since the `TIME_FMT` with `%z` was already in place historically, but since we're stabilizing the v1 of this NApp, I'll encourage you to continue this discussion, supporting timezones other than UTC with the `%z` offset fmt can really complicate things and make it harder to maintain and also to correlate events (good thing that on logs we'll have UTC on each log entry). Maybe we can nuke `pytz` and only support UTC in the v1 of this NApp? It would be great to confirm and have a discussion with network operators. In the old blueprint draft (that will be refined) on issue https://github.com/kytos-ng/maintenance/issues/59 we also didn't have any requirement to support arbitrary timezones. 

We've been using UTC everywhere on NApps and core (core also natively supports encoding datetime with this [`"%Y-%m-%dT%H:%M:%S` format](https://github.com/kytos-ng/kytos/blob/master/kytos/core/api_server.py#L33).

italovalcy replied:
Fascinating discussion, guys. I've created issue #66 to continue the discussion around this topic!

Ktmi commented:
@viniarck At this point I'm trying to hammer out the end to end tests. All I have left is 2 tests that aren't passing and I don't necessarily understand why. Could you give me a bit of insight as to why? Here is the results from running the tests: 

```
Starting enhanced syslogd: rsyslogd.
/etc/openvswitch/conf.db does not exist ... (warning).
Creating empty database /etc/openvswitch/conf.db.
Starting ovsdb-server.
Configuring Open vSwitch system IDs.
Starting ovs-vswitchd.
Enabling remote OVSDB managers.
Trying to run hello command on MongoDB...
Trying to run 'hello' command on MongoDB...
Ran 'hello' command on MongoDB successfully. It's ready!
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: rerunfailures-10.2, timeout-2.1.0
collected 24 items

tests/test_e2e_50_maintenance.py ............F......F....                [100%]

=================================== FAILURES ===================================
_____ TestE2EMaintenance.test_065_patch_mw_on_switch_new_start_delaying_mw _____

self = <tests.test_e2e_50_maintenance.TestE2EMaintenance object at 0x7fc4f49428e0>

    def test_065_patch_mw_on_switch_new_start_delaying_mw(self):
        """Tests the maintenance window data update
        process through MW's ID focused on the start time
        Test:
            /api/kytos/maintenance/v1 on POST,
            /api/kytos/maintenance/v1/{mw_id} on GET, and
            /api/kytos/maintenance/v1/{mw_id} on PATCH
        """
        self.restart_and_create_circuit()
    
        # Sets up the maintenance window information
        mw_start_delay = 30
        mw_duration = 90
        start = datetime.now() + timedelta(seconds=mw_start_delay)
        end = start + timedelta(seconds=mw_duration)
    
        # Sets up the maintenance window data
        payload = {
            "description": "mw for test 065",
            "start": start.strftime(TIME_FMT),
            "end": end.strftime(TIME_FMT),
            "switches": [
                "00:00:00:00:00:00:00:02"
            ]
        }
    
        # Creates a new maintenance window
        api_url = KYTOS_API + '/maintenance/v1'
        response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
        assert response.status_code == 201, response.text
        data = response.json()
        assert 'mw_id' in data
    
        # Extracts the maintenance window id from the JSON structure
        assert len(data) == 1
        mw_id = data["mw_id"]
    
        # Gets the maintenance schema
        api_url = KYTOS_API + '/maintenance/v1/' + mw_id
        response = requests.get(api_url)
        assert response.status_code == 200, response.text
        json_data = response.json()
        assert json_data['id'] == mw_id
    
        # Sets up a new maintenance window data
        new_start = start + timedelta(seconds=mw_start_delay)
        payload1 = {
            "start": new_start.strftime(TIME_FMT),
            "end": end.strftime(TIME_FMT),
            "switches": [
                "00:00:00:00:00:00:00:02"
            ]
        }
    
        # Updates the maintenance window information
        mw_api_url = KYTOS_API + '/maintenance/v1/' + mw_id
        request = requests.patch(mw_api_url, data=json.dumps(payload1), headers={'Content-type': 'application/json'})
        assert request.status_code == 200
    
        # Gets the maintenance window schema
        api_url = KYTOS_API + '/maintenance/v1/' + mw_id
        response = requests.get(api_url)
        json_data = response.json()
        assert json_data['start'] == new_start.strftime(TIME_FMT)
    
        # Waits for the initial MW begin time
        # (no MW, it has been changed)
        time.sleep(mw_start_delay + 5)
    
        s2 = self.net.net.get('s2')
        h11, h3 = self.net.net.get('h11', 'h3')
        h11.cmd('ip link add link %s name vlan100 type vlan id 100' % (h11.intfNames()[0]))
        h11.cmd('ip link set up vlan100')
        h11.cmd('ip addr add 100.0.0.11/24 dev vlan100')
        h3.cmd('ip link add link %s name vlan100 type vlan id 100' % (h3.intfNames()[0]))
        h3.cmd('ip link set up vlan100')
        h3.cmd('ip addr add 100.0.0.2/24 dev vlan100')
    
        # Verifies the flow at the initial MW time
        # (no maintenance at that time, it has been delayed)
        flows_s2 = s2.dpctl('dump-flows')
>       assert len(flows_s2.split('\r\n ')) == BASIC_FLOWS + 2, flows_s2
E       AssertionError:  cookie=0xac00000000000002, duration=57.752s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
E          cookie=0xac00000000000002, duration=57.746s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:01 actions=CONTROLLER:65535
E          cookie=0xab00000000000002, duration=58.459s, table=0, n_packets=40, n_bytes=1680, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
E         
E       assert 3 == (3 + 2)
E        +  where 3 = len([' cookie=0xac00000000000002, duration=57.752s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:...=58.459s, table=0, n_packets=40, n_bytes=1680, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'])
E        +    where [' cookie=0xac00000000000002, duration=57.752s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:...=58.459s, table=0, n_packets=40, n_bytes=1680, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'] = <built-in method split of str object at 0x7fc4f0f3fdc0>('\r\n ')
E        +      where <built-in method split of str object at 0x7fc4f0f3fdc0> = ' cookie=0xac00000000000002, duration=57.752s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:0...n=58.459s, table=0, n_packets=40, n_bytes=1680, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'.split

tests/test_e2e_50_maintenance.py:733: AssertionError
___________ TestE2EMaintenance.test_100_extend_running_mw_on_switch ____________

self = <tests.test_e2e_50_maintenance.TestE2EMaintenance object at 0x7fc4f494f250>

    def test_100_extend_running_mw_on_switch(self):
    
        self.restart_and_create_circuit()
    
        # Sets up the maintenance window information
        mw_start_delay = 30
        mw_duration = 60
        mw_extension = 1
        start = datetime.now() + timedelta(seconds=mw_start_delay)
        end = start + timedelta(seconds=mw_duration)
    
        # Sets up the maintenance window data
        payload = {
            "description": "mw for test 100",
            "start": start.strftime(TIME_FMT),
            "end": end.strftime(TIME_FMT),
            "switches": [
                "00:00:00:00:00:00:00:02"
            ]
        }
    
        # Creates a new maintenance window
        api_url = KYTOS_API + '/maintenance/v1'
        response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
        data = response.json()
    
        # Extracts the maintenance window id from the JSON structure
        mw_id = data["mw_id"]
    
        # Gets the maintenance schema
        api_url = KYTOS_API + '/maintenance/v1/' + mw_id
        response = requests.get(api_url)
        assert response.status_code == 200, response.text
        json_data = response.json()
        assert json_data['id'] == mw_id
    
        # Waits for the MW to start
        time.sleep(mw_start_delay + 5)
    
        # Verifies the flow behavior during the maintenance
        s2 = self.net.net.get('s2')
        flows_s2 = s2.dpctl('dump-flows')
        assert 'dl_vlan=100' not in flows_s2
        assert len(flows_s2.split('\r\n ')) == BASIC_FLOWS, flows_s2
    
        # Checks connectivity during maintenance
        h11, h3 = self.net.net.get('h11', 'h3')
        h11.cmd('ip link add link %s name vlan100 type vlan id 100' % (h11.intfNames()[0]))
        h11.cmd('ip link set up vlan100')
        h11.cmd('ip addr add 100.0.0.11/24 dev vlan100')
        h3.cmd('ip link add link %s name vlan100 type vlan id 100' % (h3.intfNames()[0]))
        h3.cmd('ip link set up vlan100')
        h3.cmd('ip addr add 100.0.0.2/24 dev vlan100')
        result = h11.cmd('ping -c1 100.0.0.2')
        assert ', 0% packet loss,' in result
    
        payload2 = {'minutes': mw_extension}
    
        # extend the maintenance window information
        api_url = KYTOS_API + '/maintenance/v1/' + mw_id + '/extend'
        response = requests.patch(api_url, data=json.dumps(payload2), headers={'Content-type': 'application/json'})
        assert response.status_code == 200, response.text
    
        # Waits to the time that the MW should be ended but instead will be running (extended)
        time.sleep(mw_duration + 5)
    
        # Verifies the flow behavior during the maintenance
        s2 = self.net.net.get('s2')
        flows_s2 = s2.dpctl('dump-flows')
        assert 'dl_vlan=100' not in flows_s2
>       assert len(flows_s2.split('\r\n ')) == BASIC_FLOWS, flows_s2
E       AssertionError:  cookie=0xac00000000000002, duration=122.910s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
E          cookie=0xac00000000000002, duration=122.874s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:01 actions=CONTROLLER:65535
E          cookie=0xaa7261a30347de40, duration=8.656s, table=0, n_packets=0, n_bytes=0, priority=20000,in_port="s2-eth2",dl_vlan=2 actions=mod_vlan_vid:2,output:"s2-eth3"
E          cookie=0xaa7261a30347de40, duration=8.647s, table=0, n_packets=1, n_bytes=78, priority=20000,in_port="s2-eth3",dl_vlan=2 actions=mod_vlan_vid:2,output:"s2-eth2"
E          cookie=0xab00000000000002, duration=123.679s, table=0, n_packets=82, n_bytes=3444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
E         
E       assert 5 == 3
E        +  where 5 = len([' cookie=0xac00000000000002, duration=122.910s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee...123.679s, table=0, n_packets=82, n_bytes=3444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'])
E        +    where [' cookie=0xac00000000000002, duration=122.910s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee...123.679s, table=0, n_packets=82, n_bytes=3444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'] = <built-in method split of str object at 0x149c620>('\r\n ')
E        +      where <built-in method split of str object at 0x149c620> = ' cookie=0xac00000000000002, duration=122.910s, table=0, n_packets=0, n_bytes=0, priority=50000,dl_src=ee:ee:ee:ee:ee:...=123.679s, table=0, n_packets=82, n_bytes=3444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535\r\n'.split

tests/test_e2e_50_maintenance.py:1076: AssertionError
=============================== warnings summary ===============================
test_e2e_50_maintenance.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1121: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    return ( StrictVersion( cls.OVSVersion ) <

test_e2e_50_maintenance.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1122: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    StrictVersion( '1.10' ) )

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
------------------------------- start/stop times -------------------------------
test_e2e_50_maintenance.py::TestE2EMaintenance::test_065_patch_mw_on_switch_new_start_delaying_mw: 2023-01-04,17:22:59.587115 - 2023-01-04,17:24:14.794649
test_e2e_50_maintenance.py::TestE2EMaintenance::test_100_extend_running_mw_on_switch: 2023-01-04,17:30:35.551109 - 2023-01-04,17:32:55.960302
=========================== short test summary info ============================
FAILED tests/test_e2e_50_maintenance.py::TestE2EMaintenance::test_065_patch_mw_on_switch_new_start_delaying_mw
FAILED tests/test_e2e_50_maintenance.py::TestE2EMaintenance::test_100_extend_running_mw_on_switch
============ 2 failed, 22 passed, 34 warnings in 1503.82s (0:25:03) ============

```

Both failures seem to link back to the amount of flows on the switch not matching the expected amount, but beyond that I don't understand the issue.

italovalcy reviewed (changes requested):
@Ktmi can you please check lint issues?

italovalcy commented on main.py:
```diff
@@ -45,118 +47,118 @@ def shutdown(self):
-            raise BadRequest('Start in the past not allowed')
-        if maintenance.end <= maintenance.start:
-            raise BadRequest('End before start not allowed')
+            maintenance = MW.parse_obj(data)
```
From what I could investigate, this issue was fixed:

```
{"code":400,"description":"Maintenance Window 'c8fec28478eb4b6e876709df2fed4509'<2023-01-13 14:00:00+00:00 to 2023-01-13 15:00:00+00:00> interferes with the following windows: ['94a37860a1704ea4ad2095493ff07bc0'<2023-01-13 14:00:00+00:00 to 2023-01-13 15:00:00+00:00>]","name":"Bad Request"}
```

italovalcy commented:
I've executed a new end-to-end test for the specific tests of the maintenance Napp and everything looks good:

```
+ python3 -m pytest tests/test_e2e_50_maintenance.py --reruns 2 -r fEr
================================================================================================= test session starts =================================================================================================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /kytos-end-to-end-tests
plugins: rerunfailures-10.2, timeout-2.1.0
collected 24 items

tests/test_e2e_50_maintenance.py ........................                                                                                                                                                       [100%]

================================================================================================== warnings summary ===================================================================================================
tests/test_e2e_50_maintenance.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1121: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    return ( StrictVersion( cls.OVSVersion ) <

tests/test_e2e_50_maintenance.py: 17 warnings
  /usr/lib/python3/dist-packages/mininet/node.py:1122: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    StrictVersion( '1.10' ) )

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
-------------------------------------------------------------------------------------------------- start/stop times ---------------------------------------------------------------------------------------------------
==================================================================================== 24 passed, 34 warnings in 1706.57s (0:28:26) =====================================================================================
```

italovalcy reviewed (approved):
LGTM. @Ktmi, congratulations and thank you very much for your contributions on this PR! It is very nice to see Maintenance getting more and more robust!
